source/utilities.tex

```diff
@@ -11849,7 +11849,7 @@
 \remarks
 Two \tcode{shared_ptr} objects are equivalent if
 they store the same pointer value and
-either share ownership, or both are empty.
+share ownership, or if both are empty.
```

Comment:
Hmm. This reads like a normative change, but appears correct: "empty" is defined as "does not own a pointer", so the previous "store the same pointer value and both are empty" case was meaningless by definition.

Reply:
This is a normative change. It's possible for a `shared_ptr` to store a pointer that it does not own, via the aliasing constructors specified in [util.smartptr.shared.const]/14 - /17. Consequently two equivalent `shared_ptr`s can:
* store the same pointer value and share ownership, or
* store the same pointer value and both be empty.

 The proposed change makes *all* empty `shared_ptr`'s equivalent, regardless of their stored pointer values.

Reply:
This is a normative change. [util.smartptr.shared.const]/17:
> [Note: This constructor allows creation of an empty `shared_ptr` instance with a non-null stored pointer.—end note]

Reply:
@CaseyCarter: I think you're right. I was on autopilot here and thinking about the conditions for two shared_ptrs to be *equal* (`operator==`-wise); but this section is talking about the proposed semantics for *equivalence* (`compare_exchange_strong`-wise).

No vendor has yet implemented `atomic<shared_ptr<T>>` as far as I can tell, but if anyone ever did, I think they'd naturally make `shared_ptr<T>(p1, nullptr)` "equal but non-equivalent to" `shared_ptr<T>(p2, nullptr)`. That is, `compare_exchange_strong` will look at the bitwise values of the control-block pointer and the value pointer and treat the shared_ptrs as equivalent only if *both* pointers are equal.

https://wandbox.org/permlink/xAepLm6DdFXG5oa9

I think I'm unqualified to propose a "correct" wording here, especially if @zygoloid is correct that the original wording is meaningless. I'll take suggestions.

Reply:
A `shared_ptr` can **store** a pointer without **owning** it, and can share one pointer and own a different pointer. The original wording is not meaningless, it's correct.

Reply:
> A `shared_ptr` can **store** a pointer without **owning** it, and can **store** one pointer and **own** a different pointer. The original wording is not meaningless, it's correct.

For completeness: a `shared_ptr` can also own an object of type `nullptr_t` via the `(nullptr_t, deleter)` and `(nullptr_t, deleter, allocator) constructors; `deleter(nullptr)` is invoked when there are no more owners. (There is more than a little weirdness in `shared_ptr`.)

Reply:
OK, so it sounds like we're settling on the original wording being correct (but maybe slightly unnaturally phrased). Thank you all for catching this! How about:

> Two `shared_ptr` objects are equivalent if they store the same pointer value, and either they share ownership or both are empty.

(With or without the comma before "and", but definitely with no comma before "or".)

Reply:
I think "Two `shared_ptr` objects are equivalent if they store the same pointer value and either share ownership or are both empty." is a bit simpler, but I certainly wouldn't object to your proposed phrasing.

Reply:
See #2793 for that change.